cranelift/codegen/src/machinst/mod.rs

```diff
@@ -269,6 +272,48 @@ pub trait MachInstLabelUse: Clone + Copy + Debug + Eq {
     fn from_reloc(reloc: Reloc, addend: Addend) -> Option<Self>;
 }
 
+/// Classification of call instruction types for granular analysis.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum CallType {
+    /// Not a call instruction.
+    None,
+    /// Regular call that returns to the caller.
+    Regular,
+    /// Tail call that doesn't return to the caller.
+    TailCall,
+}
+
+/// Enhanced function type classification for optimization purposes.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum FunctionType {
+    /// Function makes no calls at all.
+    Leaf,
+    /// Function only makes tail calls.
+    TailCallOnly,
+    /// Function has regular calls (with or without tail calls).
+    NonLeaf,
+}
+
+impl CallType {
+    /// Check if this is any kind of call (regular or tail).
+    pub fn is_call(self) -> bool {
```

Comment:
This doesn't seem to be used anywhere -- can we remove it? Best not to have methods that try to summarize/coarsen the enum domain unless we have a precise definition of what we actually want and a place that needs it; let the modeled information domain stand on its own otherwise.